lib/Dialects/LinalgTransform/Transforms/PDL.cpp

```diff
@@ -59,6 +63,138 @@ static void noOpRewriter(ArrayRef<PDLValue> args, ArrayAttr constantParams,
+    regionBuilder(b, *bodyBlock);
+  }
+
+  return haveEquivalentBodies(linalgOp, linalgModelOp, rewriter);
+}
+
+/// Check whether the unique Operation* `op` stored in `value` (assumed) is
+/// equivalent to the unique StringRefAttr operation name passed in
+/// `constantParams`.
+/// Equivalence is achieved when either:
+///   1. `op` has the name stored in `constantParams`.
+///   2. `op` and `constantParams` are both linalg ops and their structured
+///      interfaces as well as their bodies are equivalent.
+///      Structured interfaces equivalence is a simple ttribute level check.
+///      Body equivalence is more involved and currently limited:
+///        a. the current impl constructs an instance of the op whose name is
+///           specified in `constantParams` and checks for exact body equality.
+///        b. a more advanced version would "subtract" the bodies and fold, cse
+///           and canonicalize to fixed point. If the result is "all zeros",
+///           then the bodies would be equivalent (really isomorphic).
+///   3. other cases TBD (e.g. vector.generic when available).
+static LogicalResult isEquivalentToOp(PDLValue value, ArrayAttr constantParams,
+                                      PatternRewriter &rewriter) {
+  Operation *op = value.cast<Operation *>();
+  ArrayRef<Attribute> attrs = constantParams.getValue();
+  assert(attrs.size() == 1 && "expected single StringRefAttr name");
+  auto modelOpName = attrs.front().cast<StringAttr>().strref();
+
+  // 1. If op has name `modelOpName`, the match is trivial.
+  if (op->getName().getStringRef() == modelOpName)
+    return success();
+
+  // 2. Linalg vs Linalg.
+  // Create op from `constantParams`.
+  OperationState modelOpState(op->getLoc(), modelOpName, op->getOperands(),
+                              op->getResultTypes(), op->getAttrs());
+  modelOpState.addRegion();
+  Operation *modelOp = rewriter.createOperation(modelOpState);
+  auto g1 = llvm::make_scope_exit([&]() { rewriter.eraseOp(modelOp); });
+  LinalgOp linalgOp = dyn_cast<LinalgOp>(op);
+  LinalgOp linalgModelOp = dyn_cast<LinalgOp>(modelOp);
+  if (linalgOp && linalgModelOp)
+    return isEquivalentToOpImpl(linalgOp, linalgModelOp, rewriter);
+
+  // 3. TBD
+  return failure();
+}
+
+// Assume that:
+//   1. `value` is an operands range
+//   2. `constantParams` contains a DictAttr with `operand_number`, `dim` and
+//      `divisor` IntegerAttr entries.
+// Succeed if `value`[`operand_number`] is a ranked type whose `dim` is a
+// multiple of `divisor`.
+static LogicalResult isDimMultipleOf(PDLValue value, ArrayAttr constantParams,
+                                     PatternRewriter &rewriter) {
+  auto operands = value.cast<ValueRange>();
+  auto dict = constantParams.begin()->cast<DictionaryAttr>();
+  auto dim = dict.getAs<IntegerAttr>("dim").getInt();
+  auto divisor = dict.getAs<IntegerAttr>("divisor").getInt();
+  auto operandNumber = dict.getAs<IntegerAttr>("operand_number").getInt();
```

Comment:
All these look like they they can fail the typecheck and crash. Could we return failure + have debug output on why?